### PR TITLE
DOC Small documentation of multiclass classification

### PR DESCRIPTION
**Pull request by SeanMcCarren**
As suggested by @romanlutz , take the short example as shown in #752 by @hildeweerts to the user guide section on metrics with grouping.

By all means: be critical! this is my first PR :)

### CODE REVIEW

File: docs/user_guide/assessment.rst [hunk 1]
```diff
@@ -118,6 +118,30 @@ across groups and also the difference and ratio between the maximum and minimum:
     >>> print("ratio in recall = ", grouped_metric.ratio(method='between_groups'))    
     ratio in recall =  0.0
 
+We may also be interested in multiclass classification, however, typical group fairness metrics such as equalized odds and demographic parity are only defined for binary classification. We could, for every group, define one-to-one or one-to-rest classifications and calculate such metrics on this instead. However, we can also use metrics that are defined for multiclass classification. For example, accuracy is a multiclass metric, and we can easily use this using scikit-learn's :class:`sklearn.metrics.accuracy_score` in combination with a :class:`MetricFrame` as follows:
```
Comment (romanlutz) [inline docs/user_guide/assessment.rst:121]:
```suggestion
We may also be interested in multiclass classification, however, typical group fairness metrics such as equalized odds and demographic parity are only defined for binary classification. We could, for every group, define one-to-one or one-to-rest classifications and calculate such metrics on this instead. However, we can also use metrics that are defined for multiclass classification. For example, accuracy is a multiclass metric that we can utilize through scikit-learn's :class:`sklearn.metrics.accuracy_score` in combination with a :class:`MetricFrame` as follows:
```

File: docs/user_guide/assessment.rst [hunk 2]
```diff
@@ -118,6 +118,30 @@ across groups and also the difference and ratio between the maximum and minimum:
...
+    >>> from sklearn.metrics import accuracy_score
+    >>> from fairlearn.metrics import MetricFrame
+    >>> # Multiclass labels
+    >>> y_mult_true = [0,1,2,1,3,0,1,3,0,2,1,2,0,0,1,3]
+    >>> y_mult_pred = [0,1,1,2,3,0,1,0,0,2,1,2,3,0,0,2]
+    >>> # compute metric frame
+    >>> mf = MetricFrame(metric=accuracy_score,
+    ...                 y_true=y_mult_true, y_pred=y_mult_pred,
+    ...                 sensitive_features=group_membership_data)
+    >>> # print results
```
Comment (romanlutz) [inline docs/user_guide/assessment.rst:134]:
```suggestion
    >>> mf = MetricFrame(metric=accuracy_score,
    ...                  y_true=y_mult_true, y_pred=y_mult_pred,
    ...                  sensitive_features=group_membership_data)
    >>> # print results
```

I think this needs more indentation (?)

File: docs/user_guide/assessment.rst [hunk 3]
```diff
@@ -118,6 +118,30 @@ across groups and also the difference and ratio between the maximum and minimum:
...
+    >>> print(mf.by_group) # series with accuracy for each sensitive group
+    sensitive_feature_0
+    a         1.0
+    b         0.5
+    c    0.428571
+    d         1.0
+    Name: accuracy_score, dtype: object
+    >>> print(mf.difference()) # difference in accuracy between the two sensitive groups
```
Comment (romanlutz) [inline docs/user_guide/assessment.rst:142]:
```suggestion
    >>> print(mf.difference()) # difference in accuracy between the tmin and max of all groups
```

File: docs/user_guide/assessment.rst [hunk 4]
```diff
@@ -118,6 +118,30 @@ across groups and also the difference and ratio between the maximum and minimum:
...
+We may also be interested in multiclass classification, however, typical group fairness metrics such as equalized odds and demographic parity are only defined for binary classification. We could, for every group, define one-to-one or one-to-rest classifications and calculate such metrics on this instead. However, we can also use metrics that are defined for multiclass classification. For example, accuracy is a multiclass metric, and we can easily use this using scikit-learn's :class:`sklearn.metrics.accuracy_score` in combination with a :class:`MetricFrame` as follows:
+
+.. doctest:: assessment_metrics
+
+    >>> from sklearn.metrics import accuracy_score
+    >>> from fairlearn.metrics import MetricFrame
+    >>> # Multiclass labels
+    >>> y_mult_true = [0,1,2,1,3,0,1,3,0,2,1,2,0,0,1,3]
+    >>> y_mult_pred = [0,1,1,2,3,0,1,0,0,2,1,2,3,0,0,2]
+    >>> # compute metric frame
```
Comment (romanlutz) [inline docs/user_guide/assessment.rst:130]: This comment doesn't really add anything on top of the following line. I'd leave it out.

**Review by romanlutz — COMMENTED** [review]
Excellent! Thanks for making this change @SeanMcCarren!

I'm wondering whether it should be its own section or a subsection of the current section?

The FAQ page also mentions multiclass classification for assessment. It would be nice to link from there to the text you wrote so that people can read the example in detail.

In the FAQ I also saw that I spelled multclass as multi-class. What's the recommended way of spelling this? I don't particularly care except that it should be consistent.

File: docs/user_guide/assessment.rst (hunk 3 above)
Comment (SeanMcCarren) [reply docs/user_guide/assessment.rst:142]: Haha oops yes.

Comment (SeanMcCarren) [reply docs/user_guide/assessment.rst:142]: is tmin a typo? :)

Comment (romanlutz) [reply docs/user_guide/assessment.rst:142]: Yes! Fixed.

**Comment by SeanMcCarren** [comment]
> I'm wondering whether it should be its own section or a subsection of the current section?

Yeah I was wondering this too, I was thinking that this subsection (Metrics with Grouping) has a lot of topics now: 1) multiclass, 2) multiple metrics in one MetricFrame, 3) non-sample parameters, 4) Multiple sensitive features. How would you like to see this? Personally, I don't know what to think haha.

> The FAQ page also mentions multiclass classification for assessment. It would be nice to link from there to the text you wrote so that people can read the example in detail.

Smart!

> In the FAQ I also saw that I spelled multclass as multi-class. What's the recommended way of spelling this? I don't particularly care except that it should be consistent.

I'd say multiclass, as that's what I see most on scikit-learn.org, wikipedia, AIF360, but I don't know the English rules with dashes.

File: docs/user_guide/assessment.rst [hunk 5]
```diff
@@ -118,6 +118,27 @@ across groups and also the difference and ratio between the maximum and minimum:
     >>> print("ratio in recall = ", grouped_metric.ratio(method='between_groups'))    
     ratio in recall =  0.0
 
+We may also be interested in multiclass classification, however, typical group fairness metrics such as equalized odds and demographic parity are only defined for binary classification. We could, for every group, define one-to-one or one-to-rest classifications and calculate such metrics on this instead. However, we can also use metrics that are defined for multiclass classification. For example, accuracy is a multiclass metric that we can utilize through scikit-learn's :py:func:`sklearn.metrics.accuracy_score` in combination with a :class:`MetricFrame` as follows:
```
Comment (hildeweerts) [inline docs/user_guide/assessment.rst:121]:
```suggestion
We may also be interested in multiclass classification. However, typical group fairness metrics such as equalized odds and demographic parity are only defined for binary classification. One way to measure fairness in the multiclass scenario is to define one-to-one or one-to-rest classifications for each group and calculate the metrics on this instead. Alternatively, we can use predefined metrics for multiclass classification. For example, accuracy is a multiclass metric that we can utilize through scikit-learn's :py:func:`sklearn.metrics.accuracy_score` in combination with a :class:`MetricFrame` as follows:
```

**Comment by hildeweerts** [comment]
Thank you for taking a stab at this @SeanMcCarren!

> Yeah I was wondering this too, I was thinking that this subsection (Metrics with Grouping) has a lot of topics now: 1) multiclass, 2) multiple metrics in one MetricFrame, 3) non-sample parameters, 4) Multiple sensitive features. How would you like to see this? Personally, I don't know what to think haha.

I think adding subsubsections to the 'Metrics with Grouping' makes a lot of sense - that will make it much easier for people to find what they're looking for. I'm not sure though if sphinx allows that many levels in the hierarchy, though it seems possible in the [mitigation section](https://github.com/fairlearn/fairlearn/blob/main/docs/user_guide/mitigation.rst?). We might need to restructure the section a bit to make this happen.

I don't think we've been super consistent with maximum line length in the docs, but we should probably try to limit it here as well. WDYT @romanlutz?

**Comment by romanlutz** [comment]
> I don't think we've been super consistent with maximum line length in the docs, but we should probably try to limit it here as well. WDYT @romanlutz?

100%
I always try to stay below 80 (i.e., 79 max). Sometimes that's not possible (links) or desirable (math formula), but in this case it would really help.

**Comment by romanlutz** [comment]
> > In the FAQ I also saw that I spelled multclass as multi-class. What's the recommended way of spelling this? I don't particularly care except that it should be consistent.
> 
> I'd say multiclass, as that's what I see most on scikit-learn.org, wikipedia, AIF360, but I don't know the English rules with dashes.

Works for me! Do you mind adjusting the existing multi-class occurrences to be multiclass?

**Comment by SeanMcCarren** [comment]
Thanks @hildeweerts and @romanlutz for all the feedback so far :) Really appreciated!

> I think adding subsubsections to the 'Metrics with Grouping' makes a lot of sense

This is possible, but I think it makes sense to put them as separate subsections on the same level as 'Metrics with Grouping', but I'm not quite sure. Why? Because the point made is not strictly about grouped metrics for all of these subjects, I think. WDYT? ;)

**Comment by hildeweerts** [comment]
> This is possible, but I think it makes sense to put them as separate subsections on the same level as 'Metrics with Grouping', but I'm not quite sure. Why? Because the point made in these texts is not strictly about grouped metrics, I think. WDYT? ;)

All information in `MetricFrame` relate to group comparisons, which is why it is under 'Metrics with Grouping'. I can see why the headers might be a bit confusing though, as `MetricFrame.by_group` by itself is not strictly a metric on its own but a table of disaggregated "ungrouped" metrics, so some renaming/restructuring of this section is probably not a bad idea. On a related note, I'd also be in favor of retiring the Fairlearn Dashboard as a separate top-level header. This is all a bit out of the scope for this PR, though (we always try to keep PR's as small as possible to avoid blocking small (uncontroversial) changes by other changes that require more discussion). I'd suggest to open a new issue to re-evaluate the structure of the Assessment section of the user guide.

@romanlutz @SeanMcCarren

File: docs/user_guide/assessment.rst [hunk 6]
```diff
@@ -193,6 +230,9 @@ function:
     d    0...
     Name: metric, dtype: object
 
+Multiple Sensitive Features
```
Comment (romanlutz) [inline docs/user_guide/assessment.rst:233]:
```suggestion
Multiple sensitive features
```

File: docs/user_guide/assessment.rst [hunk 7]
```diff
@@ -170,6 +204,9 @@ If multiple metrics are being evaluated, then ``sample_params`` becomes a dictio
 dictionaries, with the first key corresponding matching that in the dictionary holding
 the desired underlying metric functions.
 
+Non-sample Parameters
```
Comment (romanlutz) [inline docs/user_guide/assessment.rst:207]:
```suggestion
Non-sample parameters
```

File: docs/user_guide/assessment.rst [hunk 8]
```diff
@@ -118,6 +118,40 @@ across groups and also the difference and ratio between the maximum and minimum:
...
+    sensitive_feature_0
+    a         1.0
+    b         0.5
+    c    0.428571
+    d         1.0
+    Name: accuracy_score, dtype: object
+    >>> print(mf.difference()) # difference in accuracy between the max and min of all groups
+    0.5714285714285714
+
+Multiple Metrics in one :code:`MetricFrame`
```
Comment (romanlutz) [inline docs/user_guide/assessment.rst:152]:
```suggestion
Multiple metrics in one :code:`MetricFrame`
```

File: docs/user_guide/assessment.rst [hunk 9]
```diff
@@ -118,6 +118,40 @@ across groups and also the difference and ratio between the maximum and minimum:
     >>> print("ratio in recall = ", grouped_metric.ratio(method='between_groups'))    
     ratio in recall =  0.0
 
+Multiclass Metrics
```
Comment (romanlutz) [inline docs/user_guide/assessment.rst:121]:
```suggestion
Multiclass metrics
```

File: docs/faq.rst [hunk 10]
```diff
@@ -132,14 +132,14 @@ Which ML libraries does Fairlearn support?
...
+Does Fairlearn support multiclass classification?
     On the assessment side, Fairlearn's :class:`metrics.MetricFrame`
     can be used with any metrics for supervised learning, including those for
-    multi-class classification.
+    multiclass classification.
     For example, it is possible to pass
     :py:func:`sklearn.metrics.accuracy_score` or
     :py:func:`sklearn.metrics.confusion_matrix` as the metric functions, and
-    supply multi-class data for :code:`y_true` and :code:`y_pred`.
+    supply multiclass data for :code:`y_true` and :code:`y_pred`. We give an example with multiclass data in the :ref:`user guide <metrics_with_grouping>`.
```
Comment (romanlutz) [inline docs/faq.rst:142]:
```suggestion
    supply multiclass data for :code:`y_true` and :code:`y_pred`.
    We give an example with multiclass data in the :ref:`user guide <metrics_with_grouping>`.
```

**Review by romanlutz — APPROVED** [review]
@hildeweerts +1 on opening another issue for reconsidering the structure of the assessment section.

I put a few super-nitpicky edit suggestion in the PR regarding capitalization in headers and line max lengths, but otherwise I'm fine with merging this. Thanks again @SeanMcCarren for taking this up!!!

**Comment by SeanMcCarren** [comment]
@romanlutz @hildeweerts Thanks for all the very helpful feedback!! :) Even though this was quite a trivial task, it really helps me to learn to work in this setting. 
@hildeweerts Yes, it seems smart to open a new issue for the newly discovered "issues" (the structure).

**Comment by hildeweerts** [comment]
It looks like @romanlutz's comments are marked as resolved but the suggestions are not actually committed. Was this on purpose @SeanMcCarren? It should be pretty easy to commit the suggestions right away from Github (but you first need to mark them as unresolved, I think).

**Comment by SeanMcCarren** [comment]
> 
> 
> It looks like @romanlutz's comments are marked as resolved but the suggestions are not actually committed. Was this on purpose @SeanMcCarren? It should be pretty easy to commit the suggestions right away from Github (but you first need to mark them as unresolved, I think).

Oh nice catch! I thought I changed that without using the github interface (still learning to use it haha), I'll check when I'm on my home pc.

**Comment by romanlutz** [comment]
@hildeweerts any objections to merging this? I merged the changes on `main` into this PR so that the checks can revalidate. If those pass, please merge unless you have further comments. 🙂 
Thanks again @SeanMcCarren !
